Mage/src/main/java/mage/abilities/effects/common/continuous/PlayerCanOnlyAttackInDirectionReplacementEffect.java

```diff
@@ -0,0 +1,104 @@
+    static final String ALLOW_ATTACKING_RIGHT = "Allow attacking right";
+
+    public PlayerCanOnlyAttackInDirectionReplacementEffect(Duration duration, String directionText) {
+        super(duration, Outcome.Neutral);
+        staticText = duration + (duration.toString().isEmpty() ? "" : ", ")
+                + "each player may attack only the nearest opponent "
+                + "in " + directionText + " and planeswalkers controlled by that opponent";
+    }
+
+    private PlayerCanOnlyAttackInDirectionReplacementEffect(PlayerCanOnlyAttackInDirectionReplacementEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public PlayerCanOnlyAttackInDirectionReplacementEffect copy() {
+        return new PlayerCanOnlyAttackInDirectionReplacementEffect(this);
+    }
+
+    public static Effect choiceEffect() {
+        return new ChooseModeEffect(
+                "Choose a direction to allow attacking in.",
+                ALLOW_ATTACKING_LEFT, ALLOW_ATTACKING_RIGHT
+        ).setText("choose left or right");
+    }
+
+    @Override
+    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+        return true;
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.DECLARE_ATTACKER;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        if (game.getPlayers().size() > 2) {
+            Player controller = game.getPlayer(source.getControllerId());
+            if (controller == null) {
+                return false;
+            }
+            if (!game.getState().getPlayersInRange(controller.getId(), game).contains(event.getPlayerId())) {
+                return false;
+            }
+            String allowedDirection = (String) game.getState().getValue(source.getSourceId() + "_modeChoice");
+            if (allowedDirection == null) {
+                return false;
+            }
+            Player defender = game.getPlayer(event.getTargetId());
+            if (defender == null) {
+                Permanent planeswalker = game.getPermanent(event.getTargetId());
+                if (planeswalker != null && planeswalker.getCardType(game).contains(CardType.PLANESWALKER)) {
+                    defender = game.getPlayer(planeswalker.getControllerId());
+                }
+            }
+            if (defender == null) {
+                return false;
+            }
+            PlayerList playerList = game.getState().getPlayerList(event.getPlayerId());
+            if (allowedDirection.equals(ALLOW_ATTACKING_LEFT)
+                    && !playerList.getNext().equals(defender.getId())) {
+                // the defender is not the player to the left
+                Player attacker = game.getPlayer(event.getPlayerId());
+                if (attacker != null) {
+                    game.informPlayer(attacker, "You can only attack to the left!");
```

Comment:
This is applies() method, needs to run fast for repeated checks without any GUI/choice/inform calls

Reply:
Uhm you're right. Will need to rework the whole effect.